client/transaction-pool/src/revalidation.rs

```diff
@@ -0,0 +1,312 @@
+			  self.block_ordered.remove(&first_block);
+			}
+		}
+
+		queued_exts
+	}
+
+	fn push(&mut self, worker_payload: WorkerPayload<Api>) {
+		// we don't add something that already scheduled for revalidation
+		let transactions = worker_payload.transactions;
+		let block_number = worker_payload.at;
+
+		for ext_hash in transactions {
+			// we don't add something that already scheduled for revalidation
+			if self.members.contains_key(&ext_hash) { continue; }
+
+			self.block_ordered.entry(block_number)
+				.and_modify(|value| { value.insert(ext_hash.clone()); })
+				.or_insert_with(|| {
+					let mut bt = HashSet::new();
+					bt.insert(ext_hash.clone());
+					bt
+				});
+			self.members.insert(ext_hash.clone(), block_number);
+		}
+	}
+
+	pub async fn run(mut self, from_queue: mpsc::UnboundedReceiver<WorkerPayload<Api>>) {
+		let interval = interval(BACKGROUND_REVALIDATION_INTERVAL).fuse();
+		let from_queue = from_queue.fuse();
+		futures::pin_mut!(interval, from_queue);
+		let this = &mut self;
+
+		loop {
+			futures::select! {
+				_ = interval.next() => {
+					let next_batch = this.prepare_batch();
+					batch_revalidate(this.pool.clone(), this.api.clone(), this.best_block, next_batch).await;
+				},
+				workload = from_queue.next() => {
+					match workload {
+						Some(worker_payload) => {
+							this.best_block = worker_payload.at;
+							this.push(worker_payload);
+							continue;
+						},
+						// R.I.P. worker!
+						None => break,
+					}
+				}
+			}
+		}
+	}
+}
+
+
+/// Revalidation queue.
+///
+/// Can be configured background (`new_background`)
+/// or immediate (just `new`).
+pub struct RevalidationQueue<Api: ChainApi> {
+	pool: Arc<Pool<Api>>,
+	api: Arc<Api>,
+	background: Option<mpsc::UnboundedSender<WorkerPayload<Api>>>,
+}
+
+impl<Api: ChainApi> RevalidationQueue<Api>
+where
+	Api: 'static,
+{
+	/// New revalidation queue without background worker.
+	pub fn new(api: Arc<Api>, pool: Arc<Pool<Api>>) -> Self {
+		Self {
+			api,
+			pool,
+			background: None,
+		}
+	}
+
+	/// New revalidation queue with background worker.
+	pub fn new_background(api: Arc<Api>, pool: Arc<Pool<Api>>) ->
+		(Self, Pin<Box<dyn Future<Output=()> + Send>>)
+	{
+		let (to_worker, from_queue) = mpsc::unbounded();
+
+		let worker = RevalidationWorker::new(api.clone(), pool.clone());
+
+		let queue =
+			Self {
+				api,
+				pool,
+				background: Some(to_worker),
+			};
+
+		(queue, worker.run(from_queue).boxed())
+	}
+
+	/// Queue some transaction for later revalidation.
+	///
+	/// If queue configured with background worker, this will return immediately.
+	/// If queue configured without background worker, this will resolve after
+	/// revalidation is actually done.
+	pub async fn revalidate_later(&self, at: NumberFor<Api>, transactions: Vec<ExHash<Api>>) {
+		if let Some(ref to_worker) = self.background {
+			to_worker.unbounded_send(WorkerPayload { at, transactions })
+				.expect("background task is never dropped");
```

Comment:
There is no way you can prove that from the context of this worker. The future is handed outside of this function and you have no idea what happens with it.

Reply:
It is not from the context of worker, it is the other part, which is sending.
Worker is the receiving part, and is a never-ending future.

But I agree that we can avoid panic here anyway